torch/_masked/__init__.py

```diff
@@ -512,6 +512,95 @@ def _apply(a):
     return result.coalesce()
 
 
+def _sparse_coo_scatter_reduction_helper(op,
+                                         mask_input: Tensor,
+                                         dims: Tuple[int, ...],
+                                         keepdim: bool,
+                                         dtype: Optional[DType] = None) -> Tensor:
+    reduce = op.__name__
+    assert reduce in {'sum', 'prod', 'mean', 'amax', 'amin'}, \
+        "op must be one of torch.sum, torch.prod, torch.mean, torch.amax or torch.amin"
+
+    output_dtype = dtype
+    values, indices = mask_input._values(), mask_input._indices()
+    input_dims = mask_input.dim()
+    num_sparse_dims = mask_input.sparse_dim()
+    reduced_sparse_dims = []
+    retained_sparse_dims = []
+    reduced_dense_dims = []
+
+    # promote dtype if specified
+    if values.dtype != output_dtype:
+        values = values.to(output_dtype)
+
+    if keepdim:
+        output_shape = tuple((1 if i in dims else mask_input.shape[i]) for i in range(mask_input.ndim))
+    else:
+        output_shape = tuple(mask_input.shape[i] for i in range(mask_input.ndim) if i not in dims)
+
+    for d in dims:
+        if (d >= input_dims):
+            continue
+
+        if d < num_sparse_dims:
+            reduced_sparse_dims.append(d)
+        else:
+            reduced_dense_dims.append(d + 1 - num_sparse_dims)
+
+    # Reduce dense dimensions
+    if len(reduced_dense_dims) > 0:
+        if reduce == "sum":
+            new_values = values
+            new_values = op(new_values, dim=reduced_dense_dims, keepdim=bool(keepdim))
+        else:
+            # FIXME: Implement reductions for dense dimensions for ops with non-zero reduction identities
+            raise NotImplementedError(f'{reduce}() does not support reducing dense dims of hybrid sparse coo tensors')
```

Comment:
Instead of raising a NotImplementedError it can be more convenient to return [`NotImplemented`](https://docs.python.org/3/library/constants.html?highlight=notimplemented#NotImplemented), which works better with our implementation of `__torch_dispatch__` while we're closing this gap.